File: src/main/java/me/centralhardware/znatoki/telegram/statistic/service/MinioService.java
package me.centralhardware.znatoki.telegram.statistic.service;

import com.google.common.io.Files;
import io.minio.GetObjectArgs;
import io.minio.MinioClient;
import io.minio.RemoveObjectArgs;
import io.minio.UploadObjectArgs;
import io.vavr.control.Try;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import me.centralhardware.znatoki.telegram.statistic.Config;
import org.springframework.stereotype.Component;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.InputStream;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;

@Component
@RequiredArgsConstructor
@Slf4j
public class MinioService {

    private final MinioClient minioClient;

    public Optional<String> upload(File file, LocalDateTime dateTime){
        return Try.of(() -> {
                    var fileNew = Paths.get(String.format("%s/%s/%s/%s/%s:%s-%s.jpg",
                            Config.getMinioBasePath(),
                            dateTime.getYear(),
                            dateTime.getMonth(),
                            dateTime.getDayOfMonth(),
                            dateTime.getHour(),
                            dateTime.getMinute(),
                            UUID.randomUUID()));

                    Files.createParentDirs(fileNew.toFile());
                    Files.touch(fileNew.toFile());
                    Files.move(file, fileNew.toFile());

                    minioClient.uploadObject(
                            UploadObjectArgs
                                    .builder()
                                    .bucket(Config.getMinioBucket())
                                    .filename(fileNew.toFile().getAbsolutePath())
                                    .object(fileNew.toFile().getAbsolutePath())
                                    .build()
                    );
                    //noinspection ResultOfMethodCallIgnored
                    fileNew.toFile().delete();
                    return fileNew.toFile().getAbsolutePath();
                })
                .onFailure(error -> log.warn("", error))
                .toJavaOptional();
    }

    public Try<Void> delete(String file){
        return Try.of(() -> {
            minioClient.removeObject(RemoveObjectArgs
                    .builder()
                    .bucket(Config.getMinioBucket())
                    .object(file)
                    .build());
            return null;
        });
    }

    public Try<InputStream> get(String file){
        return Try.of(() -> new ByteArrayInputStream(minioClient.getObject(GetObjectArgs
                .builder()
                .bucket(Config.getMinioBucket())
                .object(file)
                .build()).readAllBytes()));
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/service/MinioService.kt
package me.centralhardware.znatoki.telegram.statistic.service

import com.google.common.io.Files
import dev.inmo.kslog.common.KSLog
import dev.inmo.kslog.common.error
import dev.inmo.tgbotapi.requests.abstracts.asMultipartFile
import io.ktor.utils.io.core.*
import io.minio.GetObjectArgs
import io.minio.GetPresignedObjectUrlArgs
import io.minio.MinioClient
import io.minio.RemoveObjectArgs
import io.minio.UploadObjectArgs
import io.minio.http.Method
import korlibs.time.seconds
import java.io.File
import java.nio.file.Paths
import java.time.LocalDateTime
import java.util.*
import me.centralhardware.znatoki.telegram.statistic.Config
import java.util.concurrent.TimeUnit
import kotlin.time.Duration

object MinioService {

    private val minioClient =
        MinioClient.builder()
            .endpoint(Config.Minio.url)
            .credentials(Config.Minio.accessKey, Config.Minio.secretKey)
            .build()

    fun upload(file: File, dateTime: LocalDateTime): Result<String> = runCatching {
        val fileNew =
            Paths.get(
                "${Config.Minio.basePath}/${dateTime.year}/${dateTime.month}/${dateTime.dayOfMonth}/${dateTime.hour}:${dateTime.minute}=${UUID.randomUUID()}.jpg"
            )

        Files.createParentDirs(fileNew.toFile())
        Files.touch(fileNew.toFile())
        Files.move(file, fileNew.toFile())

        minioClient.uploadObject(
            UploadObjectArgs.builder()
                .bucket(Config.Minio.bucket)
                .filename(fileNew.toFile().absolutePath)
                .`object`(fileNew.toFile().absolutePath)
                .build()
        )

        fileNew.toFile().delete()
        fileNew.toFile().absolutePath
    }

    fun delete(file: String): Result<Unit> = runCatching {
        minioClient.removeObject(
            RemoveObjectArgs.builder().bucket(Config.Minio.bucket).`object`(file).build()
        )
    }

    fun get(file: String): Result<Input> = runCatching {
        minioClient
            .getObject(GetObjectArgs.builder().bucket(Config.Minio.bucket).`object`(file).build())
            .readAllBytes()
            .asMultipartFile("Отчет")
            .input
    }

    fun getLink(file: String, expire: Duration): Result<String> = runCatching {
        minioClient.getPresignedObjectUrl(
            GetPresignedObjectUrlArgs.builder()
                .method(Method.GET)
                .bucket(Config.Minio.bucket)
                .`object`(file)
                .expiry(expire.seconds.toInt(), TimeUnit.SECONDS)
                .build()
        )
    }.onFailure { KSLog.error(it) } }
